### ISSUE TITLE
Failed to get file information for file descriptor 11: Bad file descriptor

### ISSUE THREAD

**Issue by zhichengwang59** [apptainer/apptainer#1947]
### Version of Apptainer

What version of Apptainer (or Singularity) are you using? Run

1.2.5 

apptainer exec /lmod/apps/bedtools/2.31.0/bedtools_2.31.0.sif /usr/local/bin/bedtools 

### Expected behavior

should see how to use the command

### Actual behavior
ERROR  : Failed to get file information for file descriptor 11: Bad file descriptor

What actually happened? Why was it incorrect?

### Steps to reproduce this behavior
install bedtools from this Dockerfile:
FROM centos:8

LABEL \
  version="2.31.0" \
  description="bedtools2"
RUN  sed -i 's/mirrorlist/#mirrorlist/g' /etc/yum.repos.d/CentOS-*
RUN  sed -i 's|#baseurl=http://mirror.centos.org|baseurl=http://vault.centos.org|g' /etc/yum.repos.d/CentOS-*
RUN dnf -y install python3.8 python3-pip git bzip2-devel && dnf groupinstall -y "Development Tools"

WORKDIR /tmp
RUN git clone https://github.com/arq5x/bedtools2.git 
WORKDIR /tmp/bedtools2
RUN make && \
    make install
WORKDIR /
RUN rm -rf /tmp/bedtools2

then installl Posit Workbench (formaerly knwo as RStudio Server Pro)
and run the command from RStudio wbe interface

How can others reproduce this issue/problem?

### What OS/distro are you running

$ cat /etc/os-release
NAME="Rocky Linux"
VERSION="8.7 (Green Obsidian)"
ID="rocky"
ID_LIKE="rhel centos fedora"
VERSION_ID="8.7"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Rocky Linux 8.7 (Green Obsidian)"
ANSI_COLOR="0;32"
LOGO="fedora-logo-icon"
CPE_NAME="cpe:/o:rocky:rocky:8:GA"
HOME_URL="https://rockylinux.org/"
BUG_REPORT_URL="https://bugs.rockylinux.org/"
ROCKY_SUPPORT_PRODUCT="Rocky-Linux-8"
ROCKY_SUPPORT_PRODUCT_VERSION="8.7"
REDHAT_SUPPORT_PRODUCT="Rocky Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="8.7"

### How did you install Apptainer
curl -s https://raw.githubusercontent.com/apptainer/apptainer/main/tools/install-unprivileged.sh [raw.githubusercontent.com] |  bash -s - -v  1.2.5 /lmod/apps/apptainer/1.2.5/

Write here how you installed Apptainer (or Singularity). Eg. RPM, source.

**Comment by zhichengwang59** [comment]
if ssh to the server, all work OK.
we were watching the /var/log/messages while testing on the rstudio web gui. whenever that "bad file descriptor on the web interface" occurs, the overlay lowerpath message is displayed on the server .

**Comment by zhichengwang59** [comment]
here are the captured the strace output (the complete output is attached as file):
access("/lmod/apps/apptainer/1.2.2/bin/apptainer", R_OK) = 0
rt_sigprocmask(SIG_BLOCK, [INT CHLD], [], 8) = 0
lseek(255, -1, SEEK_CUR)                = 105
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x152f477bde50) = 2385409
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, [CHLD], [], 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
rt_sigprocmask(SIG_BLOCK, [CHLD], [], 8) = 0
rt_sigaction(SIGINT, {sa_handler=0x558c001f1c60, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x152f46dfeb50}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x152f46dfeb50}, 8) = 0
wait4(-1, ERROR  : Failed to get file information for file descriptor 11: Bad file descriptor
[{WIFEXITED(s) && WEXITSTATUS(s) == 1}], 0, NULL) = 2385409
rt_sigaction(SIGINT, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x152f46dfeb50}, {sa_handler=0x558c001f1c60, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x152f46dfeb50}, 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=2385409, si_uid=90023, si_status=1, si_utime=2, si_stime=6} ---
wait4(-1, 0x7ffcc1699510, WNOHANG, NULL) = -1 ECHILD (No child processes)
rt_sigreturn({mask=[]})                 = 0
read(255, "\n", 106)                    = 1
read(255, "", 106)                      = 0
rt_sigprocmask(SIG_BLOCK, [CHLD], [], 8) = 0
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
close(5)                                = 0
exit_group(1)                           = ?
+++ exited with 1 +++

**Comment by DrDaveD** [comment]
Is there any way you can reproduce it outside of Posit Workench?  It looks like installing that is an awful lot to ask someone who doesn't have it.

That strace is not helpful because it doesn't actually trace apptainer, it traces "bedtools" and does not follow to subprocesses.  A problem with using the `strace -f` option however is that typically it hangs on apptainer.  I suggest renaming your libexec/apptainer/starter command and replacing it with a shell script that runs strace on the starter, and saving that output.  The script probably needs to have full paths in it to strace and to starter.  I also recommend the `-s1024` option on strace.

**Comment by zhichengwang59** [comment]
there are no problems if we ssh to the server. everything just works fine. I have given this link to the Posit support and hopefully they can help you to install a rstudio. it is quite simple, just get the rpm and a temp license, then it should work.

**Comment by zhichengwang59** [comment]
We have resolved the “bad file descriptor” issue. it was the hyperthreading caused the problem. we have enabled and disabled logic processors  a few times and the error is reproducible.

**Comment by Marek78de** [comment]
The issue has been solved and can be closed. I summarised the issue and the resolution below. 
 
 The issue only appears on the RStudio Server (Posit Workbench) when you use Apptainer in the R Console or the R Terminal, and only if you have Hyperthreading enabled. Intermittently, the error message ERROR : Failed to get file information for file descriptor 11: Bad file descriptor will appear. It took us some time and effort to get to the bottom of it. 
 
 Thank you for your help and for looking into this!
